### PR TITLE
Design worker class

### PR DESCRIPTION
**Pull request by abcdefgs0324**
- Allow MLPWrapper to load classifier from file
- Add github util and check in github3 wrapper
- Add worker class
- Update requirements.txt

For worker class, it does
1. Subscribe events in pubsub
1. Do label prediction
1. Add labels to issues if the confidence is enough high

/assign @jlewi 
/assign @zhenghuiwang 
/assign @hamelsmu

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/code-intelligence/34)

### CODE REVIEW

File: Label-Microservice/notebooks/worker.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+        it will call the `callback` function to do label
+        prediction and add labels to the issue.
+        """
+        subscriber = pubsub.SubscriberClient()
+        subscription_path = subscriber.subscription_path(self.project_id,
+                                                         self.subscription_name)
+
+        def callback(message):
+            """
+            Object:
```
Comment (hamelsmu) [inline Label-Microservice/notebooks/worker.py]: Continue filling out the docstring

**Comment by hamelsmu** [comment]
Some questions:

- Curious why the files are in /notebooks folder
- Do you have tests or a notebook that illustrates how the MLPWrapper will be used?  I want to help you sanity check that it will be used correctly

**Comment by abcdefgs0324** [comment]
> * Curious why the files are in /notebooks folder

Oh, I made the folder disorganized. I will put all .py files to a new folder /Label-Microservice/py

> * Do you have tests or a notebook that illustrates how the MLPWrapper will be used?  I want to help you sanity check that it will be used correctly

Currently, there are only `notebook/repo_mlp.ipynb` and `tests/test_mlp.py` using MLPWrapper. I will try to add more after my current thing is done.

**Comment by hamelsmu** [comment]
I believe @jlewi proposed to move all libraries into the top level /py folder for modules.  perhaps move these modules there.  

Since this repo is supposed to share all the same dependencies, we should also potentially move the requirements.txt to the root of the repository and make sure the dependencies satisfy everyone's project.  I believe this is what ya'll decided in the debate for seperate repos  vs one monolith.

Let me know your thoughts

File: Label-Microservice/requirements.txt [hunk 2]
```diff
@@ -33,15 +33,19 @@ fastai==1.0.55
 fastprogress==0.1.21
 fire==0.1.3
 future==0.17.1
+github3.py==1.3.0
```
Comment (hamelsmu) [inline Label-Microservice/requirements.txt]: See the comment I made about Jeremy's modules at the root of /py

Having a seperate requirements.txt here and there at the same time is confusing as it is not clear which dependencies control especially if this application will rely on those general utilities.

**Comment by hamelsmu** [comment]
ok just updating this thread I talked with @abcdefgs0324 on slack and we decided to discuss the problem of repo-organization as a group.  Some tenative ideas we discussed:

 - A compromise is to remove global dependencies and have seperate dependencies per project.  This means that if you import a global utility it is that project's responsiblity to make sure there are no dependency collisions.  this is easier than maintaining global depdencies and testing them right now.
- Move all modules in py/code_intelligence into a subfolder `py/code_intelligence/notifications` to deal with ambigous files like `util.py`
- Lets add a shell scripts  folder in each application that does the following:

    - `<project_name>/script/bootstrap.sh`: prepares the environment by appending necessary items to the PYTHONPATH as well as re-building any Docker Containers as well as pushing that to a directory

    - `<project_name>/script/bootstrap.sh`: runs the application from the desired entry point  which will probably be containerized.  That way nobody has to guess the entrypoint for the application to run, this can include running the bootstrap script.

    - refernce of an example is here: https://github.com/machine-learning-apps/Issue-Label-Bot/tree/master/script

- We should try to have containers if it makes sense in each project folder.

**Comment by jlewi** [comment]
@abcdefgs0324 Doing a code refactoring (e.g. changing the directory structure) in the same PR as new code e.g. Worker class makes it hard to identify the new code changes and review them.

How about splitting this into separate PRs? 

* One PR to define a new worker class
* One PR to do any code relocation.

Regarding repo organization; maybe we can discuss in a meeting (I'll try to make it today's at least the 2nd half). Do we have examples, where the dependencies for different projects are different? i.e. where label bot depends on Version 1.0 of some library and some other project depends on Version 2.0?

If not why not just have a single requirements.txt file that specifies all requirements for all projects and organize different projects into python packages

e.g.

py/code_intelligence/label_microservice/....
py/code_intelligence/label_bot/....
py/code_intelligence/notifications/....
py/code_intelligence/util/....


* Organizing all the packages under a top level package named code_intelligence decreases the likelihood of name collisions for top level packages

    * e.g. util as a top level package would likely have name conflicts

* Organizing all top level packages under py/  is convenient for developers because they can just add a single directory to their Python path.

**Comment by jlewi** [comment]
After reviewing worker.py; here's some high level comments about repo organization

* If possible I think we want to treat this is a single python code base

   * i.e. a single requirements.txt or pip file that specifies all dependencies for all code

* Use separate python packages to distinguish various "projects" 

I think the main advantage of doing the above is that it makes it easy to reuse code across packages. For example

* Suppose we started with our label bot in py/code_intelligence/label_bot
* Suppose we define some github utilities for labeling those issues in py/code_intelligence/label_bot/github_util.py
* Now suppose we start creating a v2 of the label bot with multi-label support in py/code_intelligence/multilabel_bot
* Code in multilabel_bot can easily import `code_intelligence.label_bot.github_util` in order to reuse that code.

The problem with the original code organization is as follows

* I don't believe hyphens are valid in package names so `Label-Microservice` and `Issue-Embeddings` are not valid as python packages

  * So we need to rename these if we want to be able to treat them as python packages
  * If we don't treat them as python packages and instead do `PYTHONPATH=${PATH}/Label-Microservice` then we would likely get name conflicts because the files in those directories are not intended to be top level packages

* I think organizing all the python packages under a single top level python package named after the repo e.g. "code_intelligence.label_microservice" and "code_intelligence.issue_embeddings" is just a good way to further reduce the likelihood of top level package name conflicts

* Using "py/" to contain python code is mostly a matter of preference and a way of separating python code from other code in the repo. We could get rid of it if we don't like it.

**Comment by hamelsmu** [comment]
> Regarding repo organization; maybe we can discuss in a meeting (I'll try to make it today's at least the 2nd half). Do we have examples, where the dependencies for different projects are different? i.e. where label bot depends on Version 1.0 of some library and some other project depends on Version 2.0?

> If not why not just have a single requirements.txt file that specifies all requirements for all projects and organize different projects into python packages

@jlewi I agree about sharing code in `py`.  However, I would like to see a separate requirements.txt and Dockerfile for each project for the following reasons:

- One of the advantages of Microservices is that each person doesn't have to worry about another person's dependencies.
- We can still use common code, but we don't have to test ALL the projects each time we need to make a change to dependencies
- Example: Issue Label Bot v1 and the Embedding microservice have conflicting dependencies.  This is perfectly ok, I can still use common code across both of them but need separate requirements.txt and Dockerfiles

PTAL at https://github.com/kubeflow/code-intelligence/pull/37 I believe it makes modifications everyone agrees with

**Comment by abcdefgs0324** [comment]
Close this to split it into separate PRs in order to make it easier to be reviewed.
- Design a worker class
- Code refactoring

Will solve the above comments in the new PRs.

**Pull request closed without merging**